File: python/dsa/sorting/heap_sort.py
# Sorting — Heapsort

"""Sort an array in ascending order in place using a binary max-heap.

Technique: Heapsort
Invariant: Max-heap on active prefix; largest moved to sorted suffix.
sort: Time O(n log n); Space O(1).
"""


def sort(arr: list[int] | None) -> None:
    """Sorts arr in ascending order in place.

    Time: O(n log n); Space: O(1).
    """
    if arr is None or len(arr) < 2:
        return
    n = len(arr)
    for i in range(n // 2 - 1, -1, -1):
        _sift_down(arr, n, i)
    for end in range(n - 1, 0, -1):
        arr[0], arr[end] = arr[end], arr[0]
        _sift_down(arr, end, 0)


def _sift_down(arr: list[int], heap_size: int, i: int) -> None:
    while True:
        left = 2 * i + 1
        right = 2 * i + 2
        largest = i
        if left < heap_size and arr[left] > arr[largest]:
            largest = left
        if right < heap_size and arr[right] > arr[largest]:
            largest = right
        if largest == i:
            break
        arr[i], arr[largest] = arr[largest], arr[i]
        i = largest

File: java/src/main/java/com/dsa/examples/sorting/HeapSort.java
// Sorting — Heapsort

package com.dsa.examples.sorting;

/**
 * Sort an array in ascending order in place using a binary max-heap.
 * Technique: Heapsort
 * Invariant: Max-heap on active prefix; largest moved to sorted suffix.
 * sort: Time O(n log n); Space O(1).
 */
public final class HeapSort {

    private HeapSort() {
    }

    /**
     * Sorts {@code arr} in ascending order in place.
     * Time: O(n log n); Space: O(1).
     */
    public static void sort(int[] arr) {
        if (arr == null || arr.length < 2) {
            return;
        }
        int n = arr.length;
        for (int i = n / 2 - 1; i >= 0; i--) {
            siftDown(arr, n, i);
        }
        for (int end = n - 1; end > 0; end--) {
            swap(arr, 0, end);
            siftDown(arr, end, 0);
        }
    }

    private static void siftDown(int[] arr, int heapSize, int i) {
        while (true) {
            int left = 2 * i + 1;
            int right = 2 * i + 2;
            int largest = i;
            if (left < heapSize && arr[left] > arr[largest]) {
                largest = left;
            }
            if (right < heapSize && arr[right] > arr[largest]) {
                largest = right;
            }
            if (largest == i) {
                break;
            }
            swap(arr, i, largest);
            i = largest;
        }
    }

    private static void swap(int[] arr, int i, int j) {
        int t = arr[i];
        arr[i] = arr[j];
        arr[j] = t;
    }
}
